### ISSUE TITLE
Why GRU instead of LSTM when adding attention?

### ISSUE THREAD

**Issue by gorka96** [bentrevett/pytorch-seq2seq#47]
Hi Ben! First of all thank you for this well explained and incremental seq2seq tutorials, they have been really useful.

I was concerned, why you used LSTMs for the 1st lesson, and after that GRU layers when adding attention to the architecture. Is it just for the computational efficiency of the GRU cells, so that the tutorial runs faster, or is there any other problem to add attention to the LSTM layers?

If LSTM seq2seq models are implementable with attetion, should attention be calculated using both hidden and cell states of the LSTM encoder?   

Gorka

**Comment by bentrevett** [comment]
Thanks for the kind words.

The reason I switched from LSTM to GRU is that I was using the RNN specified in the paper the tutorial was implementing. The first used LSTMs, whilst the second two used GRUs. All tutorials should work fine with both LSTMs or GRUs as long as a few changes are made to handle the LSTMs cell state being there or not.

When using the attention-based model with an LSTM the attention should only be calculated over the hidden states and not the cell states. Thus, if you're using an LSTM for both the encoder and decoder your encoder will need to return the final cell state to be used as the initial cell state for the decoder. It should look something like this:

```python
class Encoder(nn.Module):
    def __init__(self, input_dim, emb_dim, enc_hid_dim, dec_hid_dim, dropout):
        super().__init__()
        
        self.input_dim = input_dim
        self.emb_dim = emb_dim
        self.enc_hid_dim = enc_hid_dim
        self.dec_hid_dim = dec_hid_dim
        self.dropout = dropout
        
        self.embedding = nn.Embedding(input_dim, emb_dim)
        
        self.rnn = nn.LSTM(emb_dim, enc_hid_dim, bidirectional = True)
        
        self.fc_h = nn.Linear(enc_hid_dim * 2, dec_hid_dim)
        self.fc_c = nn.Linear(enc_hid_dim * 2, dec_hid_dim)

        self.dropout = nn.Dropout(dropout)
        
    def forward(self, src):
        
        embedded = self.dropout(self.embedding(src))
              
        outputs, (hidden, cell) = self.rnn(embedded)

        hidden = torch.tanh(self.fc(torch.cat((hidden[-2,:,:], hidden[-1,:,:]), dim = 1)))
        cell = torch.tanh(self.fc_c(torch.cat((cell[-2,:,:], cell[-1,:,:]), dim = 1)))
        
        return outputs, hidden, cell
```

Your decoder will look something like this:

``` python
class Decoder(nn.Module):
    def __init__(self, output_dim, emb_dim, enc_hid_dim, dec_hid_dim, dropout, attention):
        super().__init__()

        self.emb_dim = emb_dim
        self.enc_hid_dim = enc_hid_dim
        self.dec_hid_dim = dec_hid_dim
        self.output_dim = output_dim
        self.dropout = dropout
        self.attention = attention
        
        self.embedding = nn.Embedding(output_dim, emb_dim)
        
        self.rnn = nn.LSTM((enc_hid_dim * 2) + emb_dim, dec_hid_dim)
        
        self.out = nn.Linear((enc_hid_dim * 2) + dec_hid_dim + emb_dim, output_dim)
        
        self.dropout = nn.Dropout(dropout)
        
    def forward(self, input, hidden, cell, encoder_outputs):
                
        input = input.unsqueeze(0)
                
        embedded = self.dropout(self.embedding(input))
        
        a = self.attention(hidden, encoder_outputs)
        
        a = a.unsqueeze(1)
  
        encoder_outputs = encoder_outputs.permute(1, 0, 2)
                
        weighted = torch.bmm(a, encoder_outputs)
                
        weighted = weighted.permute(1, 0, 2)
            
        rnn_input = torch.cat((embedded, weighted), dim = 2)
            
        output, (hidden, cell) = self.rnn(rnn_input, hidden.unsqueeze(0), cell.unsqueeze(0))
        
        embedded = embedded.squeeze(0)
        output = output.squeeze(0)
        weighted = weighted.squeeze(0)
        
        output = self.out(torch.cat((output, weighted, embedded), dim = 1))
               
        return output, hidden.squeeze(0), cell.squeeze(0)
```

And your Seq2Seq model will look something like this:

``` python

class Seq2Seq(nn.Module):
    def __init__(self, encoder, decoder, device):
        super().__init__()
        
        self.encoder = encoder
        self.decoder = decoder
        self.device = device
        
    def forward(self, src, trg, teacher_forcing_ratio = 0.5):
        
        batch_size = src.shape[1]
        max_len = trg.shape[0]
        trg_vocab_size = self.decoder.output_dim
        
        outputs = torch.zeros(max_len, batch_size, trg_vocab_size).to(self.device)

        encoder_outputs, hidden, cell = self.encoder(src)
                
        input = trg[0,:]
        
        for t in range(1, max_len):
            
            output, hidden, cell = self.decoder(input, hidden, cell, encoder_outputs)
            
            outputs[t] = output
            
            teacher_force = random.random() < teacher_forcing_ratio
            
            top1 = output.argmax(1) 

            input = trg[t] if teacher_force else top1

        return outputs
```

**Note**: I've just quickly written these implementations up without testing them. I don't think they have any errors but I am not 100% sure. If you try these and they work, let me know!

**Comment by gorka96** [comment]
Thanks for the fast response, I had already changed my code from GRU to LSTM, but wasn't sure if any change was needed for the attention. I will check if I implemented as you suggest here, and run it.

**Comment by gorka96** [comment]
just 2 small issues in your solution:

in the decoder, at: 
`cell = torch.tanh(self.fc_c(torch.cat((cell[-2,:,:], cell[-1,:,:]), dim = 1)))`

`fc_c` --> `fc`

and in the decoder at:
`output, (hidden, cell) = self.rnn(rnn_input, hidden.unsqueeze(0), cell.unsqueeze(0))`

`hidden.unsqueeze(0), cell.unsqueeze(0)` --> `(hidden.unsqueeze(0), cell.unsqueeze(0))`

with those changes, the code runs without problems.

**Comment by bentrevett** [comment]
@gorka96 Thanks for checking!
